### PR TITLE
Allow loading `<font>` tag styling

### LINKED ISSUE #8621: HTML elements with <font> tag cannot be loaded into CK Editor
## 📝 Provide detailed reproduction steps (if any)

1. Go to https://ckeditor.com/docs/ckeditor5/latest/examples/builds/classic-editor.html
2. In console paste
`editor.setData('<font color="#ff9900">bbb</font>');`
3. Check the HTML parsed by CK Editor

### ✔️ Expected result

The element was kept the same.
OR
The <font> tag was converted to something else, but the color information was kept.

### ❌ Actual result

Was converted to `<p>bbb</p>`. And the color was lost.

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by maxbarnas**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (font): Add `<font>` styling compatibility. Closes #8621.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-font/docs/features/font.md [hunk 1]
```diff
@@ -493,6 +493,15 @@ The {@link module:font/fontbackgroundcolor~FontBackgroundColor} plugin registers
 	We recommend using the official {@link framework/guides/development-tools#ckeditor-5-inspector CKEditor 5 inspector} for development and debugging. It will give you tons of useful information about the state of the editor such as internal data structures, selection, commands, and many more.
 </info-box>
 
+## Content compatibility
+
+The {@link module:font/font~Font} plugin provides basic support for deprecated `<font>` tag styling.
+
+While `<font color>` is always supported, to use `<font face>` and `<font size>` you need to enable {@link module:font/fontfamily~FontFamilyConfig#supportAllValues `config.fontFamily.supportAllValues`} and {@link module:font/fontsize~FontSizeConfig#supportAllValues `config.fontSize.supportAllValues`} options respectively.
```
Comment (AnnaTomanek) [inline packages/ckeditor5-font/docs/features/font.md:500]:
```suggestion
While `<font color>` is always supported, to use `<font face>` and `<font size>` you need to enable the {@link module:font/fontfamily~FontFamilyConfig#supportAllValues `config.fontFamily.supportAllValues`} and {@link module:font/fontsize~FontSizeConfig#supportAllValues `config.fontSize.supportAllValues`} options respectively.
```

File: packages/ckeditor5-font/docs/features/font.md [hunk 2]
```diff
@@ -493,6 +493,15 @@ The {@link module:font/fontbackgroundcolor~FontBackgroundColor} plugin registers
...
+
+Text formatted with `<font>` is accepted by the plugin, but the editor always returns the markup in modern format, i.e. one-way transformation.
```
Comment (AnnaTomanek) [inline packages/ckeditor5-font/docs/features/font.md:502]:
```suggestion
Text formatted with `<font>` is accepted by the plugin, but the editor always returns the markup in a modern format, so the transformation is one way only.
```

File: packages/ckeditor5-font/docs/features/font.md [hunk 3]
```diff
@@ -493,6 +493,15 @@ The {@link module:font/fontbackgroundcolor~FontBackgroundColor} plugin registers
 	We recommend using the official {@link framework/guides/development-tools#ckeditor-5-inspector CKEditor 5 inspector} for development and debugging. It will give you tons of useful information about the state of the editor such as internal data structures, selection, commands, and many more.
 </info-box>
 
+## Content compatibility
+
+The {@link module:font/font~Font} plugin provides basic support for deprecated `<font>` tag styling.
```
Comment (AnnaTomanek) [inline packages/ckeditor5-font/docs/features/font.md:498]:
```suggestion
The {@link module:font/font~Font} plugin provides basic support for the deprecated `<font>` tag.
```

File: packages/ckeditor5-font/src/fontfamily/fontfamilyediting.js [hunk 4]
```diff
@@ -110,4 +111,26 @@ export default class FontFamilyEditing extends Plugin {
 			}
 		} );
 	}
+
+	/**
+	 * Support `<font size="..">` formatting.
```
Comment (AnnaTomanek) [inline packages/ckeditor5-font/src/fontfamily/fontfamilyediting.js:116]:
```suggestion
	 * Adds support for `<font size="..">` formatting.
```

File: packages/ckeditor5-font/src/fontsize/fontsizeediting.js [hunk 5]
```diff
@@ -138,4 +151,31 @@ export default class FontSizeEditing extends Plugin {
 			}
 		} );
 	}
+
+	/**
+	 * Support `<font size="..">` formatting.
```
Comment (AnnaTomanek) [inline packages/ckeditor5-font/src/fontsize/fontsizeediting.js:156]:
```suggestion
	 * Adds support for `<font size="..">` formatting.
```

File: packages/ckeditor5-font/docs/features/font.md (hunk 2 above)
Comment (maxbarnas) [reply packages/ckeditor5-font/docs/features/font.md:502]: Thanks! That `i.e.` felt weird :)

**Comment by pomek** [comment]
> Fix (font): Add <font> styling compatibility. Closes #8621.

What does it fix? It's a feature, we didn't support the `<font>` element before.

**Comment by maxbarnas** [comment]
> > Fix (font): Add  styling compatibility. Closes #8621.
> 
> What does it fix? It's a feature, we didn't support the `<font>` element before.

The original task was marked as a bug, therefore I went with fix.

**Comment by pomek** [comment]
The Font feature wasn't designed to handle the `<font>` element so, from our point of view, it isn't a bug. But maybe we need help here.

**Comment by maxbarnas** [comment]
> The Font feature wasn't designed to handle the `<font>` element so, from our point of view, it isn't a bug. But maybe we need help here.

Happy to make it your way. Fixing a bug was a welcome side-effect, but the code here is actually a new feature.

File: packages/ckeditor5-font/tests/fontfamily/fontfamilyediting.js [hunk 6]
```diff
@@ -140,6 +140,22 @@ describe( 'FontFamilyEditing', () => {
 
 					expect( editor.getData() ).to.equal( '<p>f<span style="font-family:Arial, sans-serif;">o</span>o</p>' );
 				} );
+
+				it( 'should support <font size=".."> styling', () => {
```
Comment (niegowski) [inline packages/ckeditor5-font/tests/fontfamily/fontfamilyediting.js:144]:
```suggestion
				it( 'should support <font face=".."> styling', () => {
```

File: packages/ckeditor5-font/tests/fontfamily/fontfamilyediting.js [hunk 7]
```diff
@@ -289,5 +305,19 @@ describe( 'FontFamilyEditing', () => {
 				'<p>b<span class="text-complex">a</span>z</p>'
 			);
 		} );
+
+		it( 'should strip <font size=".."> styling when supportAllValues = false', () => {
```
Comment (niegowski) [inline packages/ckeditor5-font/tests/fontfamily/fontfamilyediting.js:309]:
```suggestion
		it( 'should strip <font face=".."> styling when supportAllValues = false', () => {
```

File: packages/ckeditor5-font/tests/fontsize/fontsizeediting.js [hunk 8]
```diff
@@ -121,6 +121,48 @@ describe( 'FontSizeEditing', () => {
 
 					expect( editor.getData() ).to.equal( '<p>f<span style="font-size:18px;">o</span>o</p>' );
 				} );
+
+				it( 'should convert <font size=".."> styling', () => {
```
Comment (niegowski) [inline packages/ckeditor5-font/tests/fontsize/fontsizeediting.js:125]: Maybe an additional test for stripping `<font size>` when `supportAllValues` is not set would be nice.

**Review by niegowski — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-font/src/fontcolor/fontcolorediting.js [hunk 9]
```diff
@@ -122,6 +122,20 @@ export default class FontColorEditing extends Plugin {
 
 		editor.commands.add( FONT_COLOR, new FontColorCommand( editor ) );
 
+		// Support `<font color="..">` formatting.
+		editor.conversion.for( 'upcast' ).elementToAttribute( {
+			view: {
+				name: 'font',
+				attributes: {
+					'color': /^#[0-9A-Fa-f]{6}|\w+$/
```
[the remark this answers is not in the document]
Comment (maxbarnas) [reply packages/ckeditor5-font/src/fontcolor/fontcolorediting.js:130]: I was going with the definition on the MDN, which was explicitly stating that color can be #RRGGBB. Happy to change that :)

File: packages/ckeditor5-font/src/fontsize/fontsizeediting.js [hunk 10]
```diff
@@ -138,4 +151,31 @@ export default class FontSizeEditing extends Plugin {
...
+	 */
+	_prepareCompatibilityConverter() {
+		const editor = this.editor;
+
+		editor.conversion.for( 'upcast' ).elementToAttribute( {
+			view: {
+				name: 'font',
+				attributes: {
+					// Size from 0 to 7.
+					'size': /^[0-7]$/
```
[the remark this answers is not in the document]
Comment (maxbarnas) [reply packages/ckeditor5-font/src/fontsize/fontsizeediting.js:168]: > Shouldn't we accept a relative `+/-` values and apply those on the default value of `3`? So for `+1` we could handle it as `4`, for `-2` we would map it to `1` etc.

The first version did support relative sizes. After discussion with @Reinmar we went with a simpler version - if I recall correctly, to mimic the old implementation from v4.

> Also, this attribute can have much bigger numbers that should be clamped to the supported range.

Same as above.

File: packages/ckeditor5-font/src/fontcolor/fontcolorediting.js [hunk 11]
```diff
@@ -122,6 +122,20 @@ export default class FontColorEditing extends Plugin {
 
 		editor.commands.add( FONT_COLOR, new FontColorCommand( editor ) );
 
+		// Support `<font color="..">` formatting.
+		editor.conversion.for( 'upcast' ).elementToAttribute( {
```
[the remark this answers is not in the document]
Comment (maxbarnas) [reply packages/ckeditor5-font/src/fontcolor/fontcolorediting.js:126]: Went with adding it at the bottom to visually separate it from "regular" conversion, but I am totally okay with having them together.

File: packages/ckeditor5-font/src/fontsize/fontsizeediting.js [hunk 12]
```diff
@@ -14,6 +14,18 @@ import FontSizeCommand from './fontsizecommand';
 import { normalizeOptions } from './utils';
 import { buildDefinition, FONT_SIZE } from '../utils';
 
+// Mapping of `<font size="..">` styling to CSS's `font-size` values.
+const styleFontSize = [
+	'x-small', // Size "0" equal to "1".
```
[the remark this answers is not in the document]
Comment (maxbarnas) [reply packages/ckeditor5-font/src/fontsize/fontsizeediting.js:19]: "Thanks" to regex I did not have to care about really weird values. That's why it is so limited here. If we change the conversion regex, then I agree about supporting bigger values.

File: packages/ckeditor5-font/tests/fontsize/fontsizeediting.js (hunk 8 above)
Comment (maxbarnas) [reply packages/ckeditor5-font/tests/fontsize/fontsizeediting.js:125]: Sure, just added.

File: packages/ckeditor5-font/src/fontcolor/fontcolorediting.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (maxbarnas) [reply packages/ckeditor5-font/src/fontcolor/fontcolorediting.js:119]: Moved it closer together. Unfortunately `upcastAlso` is not flexible enough to use here.

File: packages/ckeditor5-font/src/fontsize/fontsizeediting.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (maxbarnas) [reply packages/ckeditor5-font/src/fontsize/fontsizeediting.js:19]: Fixed. Made it handle values up to 999.

File: packages/ckeditor5-font/src/fontsize/fontsizeediting.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (maxbarnas) [reply packages/ckeditor5-font/src/fontsize/fontsizeediting.js]: Done. Made it handle values between -999 and +999.

File: packages/ckeditor5-font/src/fontcolor/fontcolorediting.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (maxbarnas) [reply packages/ckeditor5-font/src/fontcolor/fontcolorediting.js]: Relaxed the regex and let it work in a similar manner as the regular non-legacy color upcaster - which is pretty much passing down the values unchanged.

File: packages/ckeditor5-font/src/fontsize/fontsizeediting.js [hunk 13]
```diff
@@ -138,4 +151,43 @@ export default class FontSizeEditing extends Plugin {
...
+					'size': /^[+-]?\d{1,3}$/
+				}
+			},
+			model: {
+				key: FONT_SIZE,
+				value: viewElement => {
+					const value = viewElement.getAttribute( 'size' );
+					const isRelative = value[ 0 ] === '-' || value[ 0 ] === '+';
+
+					let size = Number( value );
```
Comment (niegowski) [inline packages/ckeditor5-font/src/fontsize/fontsizeediting.js:178]: I would suggest using `parseInt( value, 10 )`.

File: packages/ckeditor5-font/src/fontcolor/fontcolorediting.js [hunk 14]
```diff
@@ -115,6 +115,20 @@ export default class FontColorEditing extends Plugin {
...
+			view: {
+				name: 'font',
+				attributes: {
+					'color': /^#?\w+$/
+				}
+			},
+			model: {
+				key: FONT_COLOR,
+				value: viewElement => viewElement.getAttribute( 'color' )
```
Comment (niegowski) [inline packages/ckeditor5-font/src/fontcolor/fontcolorediting.js:128]: For the `AABBCC` format that would not be a valid color but it is handled properly by the browsers in the `<font>` tag but not in CSS. But I think that information won't be lost if we don't fix it (it's still in the data but in invalid format).

File: packages/ckeditor5-font/tests/fontsize/fontsizeediting.js [hunk 15]
```diff
@@ -121,6 +121,121 @@ describe( 'FontSizeEditing', () => {
...
+						'<paragraph><$text fontSize="x-small">[]foo</$text><$text fontSize="18px">bar</$text></paragraph>'
+					);
+
+					expect( editor.getData() ).to.equal(
+						'<p><span style="font-size:x-small;">foo</span><span style="font-size:18px;">bar</span></p>'
+					);
+				} );
+
+				// TODO: handle big sizes 		[ ]
+				// TODO: handle relative sizing	[ ]
```
Comment (niegowski) [inline packages/ckeditor5-font/tests/fontsize/fontsizeediting.js:154]: Please remove this 😉

File: packages/ckeditor5-font/src/fontsize/fontsizeediting.js (hunk 13 above)
Comment (maxbarnas) [reply packages/ckeditor5-font/src/fontsize/fontsizeediting.js:178]: Done.

File: packages/ckeditor5-font/tests/fontsize/fontsizeediting.js (hunk 15 above)
Comment (maxbarnas) [reply packages/ckeditor5-font/tests/fontsize/fontsizeediting.js:154]: 😉

**Comment by Mgsy** [comment]
PR is ready for QA phase, @FilipTokarski @LukaszGudel please take a look at it.

**Comment by LukaszGudel** [comment]
LGTM :+1:

**Comment by FilipTokarski** [comment]
Looks ok, I didn't find any bugs 👍

**Pull request merged**